### PR TITLE
MM-T338 Image Attachment Upload in Mobile View, MM-T345 Public links of common file types should open in a new tab

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
This PR adds end to end tests to codebase which tries to verify if image attachments in mobile view are correctly rendered pre and post sending of message.

It also adds an test to check if public links of common file types opens in a new tab with their oreviews

#### Ticket Link
MM-T338
MM-T345

#### Related Pull Requests
none

#### Screenshots
none

### CODE REVIEW

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 1]
```diff
@@ -179,20 +189,134 @@ describe('Upload Files', () => {
...
+        cy.findByLabelText('message input complimentary region').should('be.visible').within(() => {
+            // * Check if the attachment button is present
+            cy.findByLabelText('Attachment Icon').should('be.visible').and('have.css', 'cursor', 'pointer');
+        });
+
+        const imageName = 'image-file';
+        const imageType = 'jpg';
+        const imageFilename = `${imageName}.${imageType}`;
+
+        // # Attach an image but dont post it yet
```
Comment (saturninoabril) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:207]:
```suggestion
        // # Attach an image but don't post it yet
```

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 2]
```diff
@@ -179,20 +189,134 @@ describe('Upload Files', () => {
...
+    cy.request(fileURL).then((response) => {
         // * Verify that link can be downloaded
         expect(response.status).to.equal(200);
 
-        // * Verify if link is an attachment that can be saved locally
+        // * Verify if link is an appropriate httpContext for opening in new tab or same and that can be saved locally
         // and it contains the correct filename* which will be used to name the downloaded file
         expect(response.headers['content-disposition']).to.
-            equal(`attachment;filename="${filename}"; filename*=UTF-8''${filename}`);
+            equal(`${httpContext};filename="${filename}"; filename*=UTF-8''${filename}`);
```
Comment (saturninoabril) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:320]: Could you confirm if all are passing on you? I've got failing test here.  See screenshot.
Screen Shot 2020-09-01 at 6 34 20 PM

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 3]
```diff
@@ -179,20 +189,134 @@ describe('Upload Files', () => {
...
+                // * Check that image is in the post message with valid source link
+                cy.findByLabelText(`file thumbnail ${imageFilename}`).should('be.visible').
+                    and('have.attr', 'src').then((src) => {
+                        downloadAttachmentAndVerifyItsProperties(src, imageFilename, 'inline');
+                    });
+            });
+        });
+    });
+
+    it.only('MM-T345 Public links for common file types should open in a new browser tab', () => {
```
Comment (saturninoabril) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:244]:
```suggestion
    it('MM-T345 Public links for common file types should open in a new browser tab', () => {
```

File: e2e/cypress/support/ui_commands.js [hunk 4]
```diff
@@ -91,6 +91,16 @@ Cypress.Commands.add('postMessageReplyInRHS', (message) => {
     postMessageAndWait('#reply_textbox', message);
 });
 
+Cypress.Commands.add('postMessageQuickly', (message) => {
```
Comment (saturninoabril) [inline e2e/cypress/support/ui_commands.js:94]: Please change to `uiPostMessageQuickly` to easily check at a glance that this is done via UI. It's our recent convention for UI commands, although not many were migrated yet. Also, please add types at `ui_commands.d.ts`.  Thanks!

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks @M-ZubairAhmed.  Please see comments.

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 5]
```diff
@@ -149,26 +159,32 @@ describe('Upload Files', () => {
...
 
     it('MM-T341 Download link on preview - Image file (non SVG)', () => {
-        ['small-image.png', 'image-file.jpg', 'image-file.bmp', 'image-file.gif', 'image-file.tiff'].forEach((image) => {
-            cy.get('#centerChannelFooter').find('#fileUploadInput').attachFile(image);
-            cy.get('.post-image').should('be.visible');
-            cy.postMessage(MESSAGES.SMALL);
-            cy.uiWaitUntilMessagePostedIncludes(MESSAGES.SMALL);
-            cy.getLastPostId().then((postId) => {
-                cy.get(`#post_${postId}`).within(() => {
+        ['bmp-image-file.bmp', 'png-image-file.png', 'jpg-image-file.jpg', 'gif-image-file.gif', 'tiff-image-file.tif'].
```
Comment (M-ZubairAhmed) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:162]: This particular tests bought in its own set of images, Images as per the docs (@saturninoabril google drive) were provided in this PR, i replaced those in this test and removed now redundant images from fixtures

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 6]
```diff
@@ -149,26 +159,32 @@ describe('Upload Files', () => {
...
+                        cy.get('div.image-loaded').find('img').click();
+                    });
                 });
-            });
-            cy.get('.a11y__modal').should('be.visible').within(() => {
+                cy.get('.a11y__modal').should('be.visible').within(() => {
                 // # Click the "Download" link in the previewer
-                cy.findByText('Download').should('be.visible').parent().then((fileAttachment) => {
-                    //* Image should download
-                    verifyLinkHasDownloadProperties(fileAttachment, image);
```
Comment (M-ZubairAhmed) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:167]: `verifyLinkHasDownloadProperties` was changed and improved in this PR, so made the change

**Review by prapti — CHANGES_REQUESTED** [review]
Thank you for working on these tests @M-ZubairAhmed! Great improvements on things that affect the previous tests you've worked on :) 

A couple of comments and questions before I can approve it:
1. I think it's a good idea to separate both the tests for MM-T338 and MM-T345 in new spec files each.
2. It looks like MM-T345 is missing some test steps as mentioned in the [test case](https://automation-test-cases.vercel.app/test/MM-T345), esp. steps 6 (Click on "Copy Link"), and 8 (Click on the link to verify that the file opens in a new browser tab), unless I'm missing something? Please let me know if I've overlooked any details. Thanks!

**Comment by saturninoabril** [comment]
@M-ZubairAhmed Let us know status of this PR, thanks!

**Comment by M-ZubairAhmed** [comment]
Hi @prapti , thank you for your review
1. I have a contrary opinion on that, since all of them belongs to some form of upload, i have placed them in the same file. They both also share the same `downloadAttachmentAndVerifyItsProperties` function.
2. Here are the steps with respect to code line numbers
- Account Settings ➜ Display ➜ Image Previews ➜ Expanded : https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R358
- Make a post with an attachment of any of the following types jpeg, png, bmp, .tiff, gif, avi, mpeg, mp4, mpeg, or wav: https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R377
- Click on the image preview : https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R384
- Click on "Get public link" : https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R392
- Click on "Copy link", then on "Close" : https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R402 & https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R409
- Paste the link into the message input box and hit "Enter" to make the post : https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R415
- Click on the link: https://github.com/mattermost/mattermost-webapp/pull/6321/files#diff-c9a96f255e269faf4adce4e32b2b0bb7R418

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 7]
```diff
@@ -309,20 +325,133 @@ describe('Upload Files', () => {
         cy.postMessage('test');
         cy.findByTestId('fileAttachmentList').find('.post-image').should('have.length', 5);
     });
-});
```
Comment (saturninoabril) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:312]: With the changes on files, L233 failed on me when running all tests in this spec. It should now be:
```diff
- const attachmentFilename = 'image-file.jpg';
+ const attachmentFilename = 'jpg-image-file.jpg';
```

Please run all the tests and let me know if all are consistently passing on you.

File: e2e/cypress/integration/files_and_attachments/upload_files_spec.js [hunk 8]
```diff
@@ -309,20 +325,133 @@ describe('Upload Files', () => {
...
+            cy.findByText(imageFilename).should('be.visible');
+
+            // * Verify that image type is present
+            cy.findByText(imageType).should('be.visible');
+
+            // # Get the image preview div
+            cy.get('.post-image.normal').then((imageDiv) => {
+                // # Filter out the url from the css background property
+                // url("https://imageurl") => https://imageurl
+                const imageURL = imageDiv.css('background-image').replace('url(', '').replace(')', '').replace(/"/gi, '');
```
Comment (saturninoabril) [inline e2e/cypress/integration/files_and_attachments/upload_files_spec.js:357]: I suggests,
```suggestion
                const imageURL = imageDiv.css('background-image').split('"')[1];
```

**Review by saturninoabril — CHANGES_REQUESTED** [review]
Thanks @M-ZubairAhmed! Overall looks good to me, except for last couple comments.

**Comment by saturninoabril** [comment]
Hello Zubair, let us know once ready for review.  Thanks!

**Comment by M-ZubairAhmed** [comment]
@saturninoabril i made the changes, i ran them individually. Can you please run them collectively?

**Review by saturninoabril — APPROVED** [review]
Thanks @M-ZubairAhmed! Tested and all passed. LGTM.

**Review by prapti — APPROVED** [review]
LGTM. Thank you for the improvements @M-ZubairAhmed!

**Pull request merged**